File: sim_ws/src/ar_tag_sim/src/axis_corrector.py
#!/usr/bin/env python 

# Halil Faruk Karagoz 
# At first i wrote this for fixing the axis then it was working correct than we fixed it with static frame 
#After this code wasn't necessary to use but still couldn't delete it bcs it was my first proper ros code <kiyamadim>
import rospy 
import tf2_geometry_msgs
import tf
from geometry_msgs.msg import Transform
from math import pi
from fiducial_msgs.msg import FiducialTransformArray,FiducialTransform


class axis_correction:
    def __init__(self):
        self.correct_transforms = FiducialTransformArray()
        rospy.init_node("axis_corrector")
        self.listener = tf.TransformListener()
        self.br = tf.TransformBroadcaster()
        rospy.Subscriber("/fiducial_transforms",FiducialTransformArray,self.fiducial_correct)
        self.pub = rospy.Publisher("/fiducial_correct",FiducialTransform,queue_size=10)
        rospy.spin()


    def fiducial_correct(self,data):
        self.correct_transforms.header = data.header
        self.correct_transforms.image_seq = data.image_seq
        for i in data.transforms:
            self.correct_transform = FiducialTransform()

            self.correct_transform.fiducial_id = i.fiducial_id
            self.correct_transform.image_error = i.image_error
            self.correct_transform.object_error = i.object_error
            self.correct_transform.fiducial_area = i.fiducial_area


            self.correct_transform.transform.translation.x = i.transform.translation.z
            self.correct_transform.transform.translation.y = -i.transform.translation.x
            self.correct_transform.transform.translation.z = -i.transform.translation.y
            
            self.roll,self.pitch,self.yaw = tf.transformations.euler_from_quaternion([i.transform.rotation.x,i.transform.rotation.y,i.transform.rotation.z,i.transform.rotation.w])
            x,y,z,w = tf.transformations.quaternion_from_euler(self.yaw,-self.roll + pi/2,-self.pitch)
            self.correct_transform.transform.rotation.x = x
            self.correct_transform.transform.rotation.y = y
            self.correct_transform.transform.rotation.z = z
            self.correct_transform.transform.rotation.w = w
            self.pub.publish(self.correct_transform)
            transform = (self.correct_transform.transform.translation.x,self.correct_transform.transform.translation.y,self.correct_transform.transform.translation.z)
            rotation = (self.correct_transform.transform.rotation.x,self.correct_transform.transform.rotation.y,self.correct_transform.transform.rotation.z,self.correct_transform.transform.rotation.w)

            self.br.sendTransform(transform,
                        rotation,
                        rospy.Time.now(),
                        str(self.correct_transform.fiducial_id),
                        "camera"
                        )
            #print(type(("ar_marker %d",data.id)))
if __name__ == "__main__":
    axis_correction()





    

File: sim_ws/src/ar_tag_sim/src/axis_corrector.cpp
#include "ros/ros.h"
#include "tf/tf.h"
#include "visualization_msgs/Marker.h"
#include "geometry_msgs/Quaternion.h"
#include <iostream>
#include <cmath>

ros::Publisher pub;

void vis_marker_correct(const visualization_msgs::Marker::ConstPtr& data){
    visualization_msgs :: Marker correct_marker;
    correct_marker = *(data);
    correct_marker.pose.position.x = data->pose.position.z;
    correct_marker.pose.position.y = -data->pose.position.x;
    correct_marker.pose.position.z = data->pose.position.y;
    tf::Quaternion q(
        data->pose.orientation.x,
        data->pose.orientation.y,
        data->pose.orientation.z,
        data->pose.orientation.w);
    tf::Matrix3x3 m(q);
    double roll, pitch, yaw;
    m.getRPY(roll, pitch, yaw);
    tf::Quaternion correct_q;
    correct_q.setRPY(yaw,roll + M_PI_2,-pitch);
    correct_marker.pose.orientation.x = correct_q.getX();
    correct_marker.pose.orientation.y = correct_q.getY();
    correct_marker.pose.orientation.z = correct_q.getZ();
    correct_marker.pose.orientation.w = correct_q.getW();
    pub.publish(correct_marker);
}


int main(int argc, char **argv){
    ros::init(argc,argv,"axis_corrector");
    ros::NodeHandle n;
    ros::Rate rate(100);
    pub = n.advertise<visualization_msgs::Marker>("vis_marker_correct",100);
    ros::Subscriber sub  =n.subscribe("visualization_marker", 100 ,vis_marker_correct);
    ros :: spin();

    return 0;
}